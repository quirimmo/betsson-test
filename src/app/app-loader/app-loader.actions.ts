import { Injectable } from '@angular/core';
import { Action } from 'redux';
import { NgRedux } from '@angular-redux/store';
import { StoreModel } from '../store/store.model';

export interface AppLoaderAction extends Action {
	isLoading: boolean;
}

@Injectable()
export class AppLoaderActions {
	static readonly START_LOADING = 'START_LOADING';
	static readonly STOP_LOADING = 'STOP_LOADING';

	constructor(private ngRedux: NgRedux<StoreModel>) {}

	public startLoading(): AppLoaderAction {
		return this.ngRedux.dispatch({
			type: AppLoaderActions.START_LOADING,
			isLoading: true
		});
	}

	public stopLoading(): AppLoaderAction {
		return this.ngRedux.dispatch({
			type: AppLoaderActions.STOP_LOADING,
			isLoading: false
		});
	}
}
