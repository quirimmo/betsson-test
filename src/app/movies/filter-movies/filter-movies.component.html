<form>
	<div class="row search-movie-container">
		<mat-form-field class="col-md-6 col-12 w-100">
			<input
				id="search-movie-name"
				[formControl]="nameText"
				matInput
				placeholder="Type the name of the movie..."
			/>
		</mat-form-field>
		<mat-form-field class="col-md-6 col-12 w-100">
			<mat-select
				id="seach-movie-genres"
				placeholder="Select the genres..."
				[formControl]="genresSelect"
				multiple
			>
				<mat-option *ngFor="let genre of genres" [value]="genre">{{
					genre
				}}</mat-option>
			</mat-select>
		</mat-form-field>
	</div>
</form>
