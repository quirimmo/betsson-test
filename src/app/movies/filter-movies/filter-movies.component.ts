import { Component, Input, OnInit, OnDestroy } from '@angular/core';
import { FormControl } from '@angular/forms';
import { Subscription } from 'rxjs';
import { ActivatedRoute, Router } from '@angular/router';
import { MoviesRouterService } from '../movies-router.service';

@Component({
	selector: 'filter-movies',
	inputs: ['genres'],
	styleUrls: ['./filter-movies.component.scss'],
	templateUrl: './filter-movies.component.html'
})
export class FilterMoviesComponent implements OnInit, OnDestroy {
	@Input()
	genres: string[];
	genresSelect = new FormControl();
	nameText = new FormControl();
	nameTextChangesSubscriber: Subscription;
	genresChangesSubscriber: Subscription;

	constructor(
		private moviesRouterService: MoviesRouterService,
		private route: ActivatedRoute,
		private router: Router
	) {}

	ngOnInit(): void {
		const searchName: string =
			this.moviesRouterService.getSearchNameParam() || '';
		const searchGenres: string[] =
			this.moviesRouterService.getSearchGenresParam() || this.genres;

		this.nameTextChangesSubscriber = this.nameText.valueChanges.subscribe(
			this.onChangeSearchName.bind(this)
		);
		this.genresChangesSubscriber = this.genresSelect.valueChanges.subscribe(
			this.onChangeSearchGenres.bind(this)
		);

		this.nameText.setValue(searchName);
		this.genresSelect.setValue(searchGenres);
	}

	ngOnDestroy(): void {
		this.nameTextChangesSubscriber.unsubscribe();
		this.genresChangesSubscriber.unsubscribe();
	}

	public onChangeSearchName(): void {
		this.moviesRouterService.appendSearchNameParam(this.nameText.value);
	}

	public onChangeSearchGenres(): void {
		this.moviesRouterService.appendSearchGenresParam(this.genresSelect.value);
	}
}
