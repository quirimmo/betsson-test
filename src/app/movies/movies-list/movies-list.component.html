<section>
	<section>
		<filter-movies #filterMovies [genres]="genres"></filter-movies>
	</section>
	<section class="row">
		<div
			*ngIf="
				isNoMoviesMessageVisible(
					filterMovies.nameText,
					filterMovies.genresSelect
				)
			"
			class="col-12 justify-content-center"
			[@slideInLeftOnEnter]
			[@slideOutRightOnLeave]
		>
			<p class="no-movies-list-message">
				It looks like there are not movies corresponding to your search...
				<br />
				Try changing your filters!
			</p>
		</div>
		<div
			*ngIf="
				!isMoviesListEmpty(
					filterMovies.nameText.value,
					filterMovies.genresSelect.value
				)
			"
			class="col-12"
		>
			<div class="row">
				<section class="col-12 mb-5">
					<sort-movies #sortMovies></sort-movies>
				</section>
				<article
					class="col-xl-2 col-lg-3 col-md-4 col-sm-6 col-12 movie-list-item-wrapper"
					*ngFor="
						let movie of (getFilteredMovies(
							filterMovies.nameText.value,
							filterMovies.genresSelect.value
						) | sortMovies: sortMovies.sortBySelect.value);
						trackBy: trackById
					"
					[@bounceInUpOnEnter]
					[@bounceOutOnLeave]
				>
					<a [routerLink]="['/movie', movie.id]">
						<movie-list-item [movie]="movie"></movie-list-item>
					</a>
				</article>
			</div>
		</div>
	</section>
</section>
