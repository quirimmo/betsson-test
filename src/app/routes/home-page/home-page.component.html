<section class="container-fluid">
	<movies-list [movies]="movies"> </movies-list>
</section>
