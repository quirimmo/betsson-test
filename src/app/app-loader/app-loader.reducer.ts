import { AppLoaderAction, AppLoaderActions } from './app-loader.actions';

export function appLoaderReducer(
	state: boolean = false,
	action: AppLoaderAction
): boolean {
	switch (action.type) {
		case AppLoaderActions.START_LOADING:
			return action.isLoading;
		case AppLoaderActions.STOP_LOADING:
			return action.isLoading;
		default:
			return state;
	}
}
