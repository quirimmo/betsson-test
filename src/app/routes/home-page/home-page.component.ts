import { Component, OnInit } from '@angular/core';
import { AppLoaderActions } from '../../app-loader/app-loader.actions';
import Movie from 'src/app/movies/movie.model';
import { ActivatedRoute } from '@angular/router';

@Component({
	selector: 'home-page',
	templateUrl: './home-page.component.html'
})
export class HomePageComponent implements OnInit {
	movies: Movie[] = [];

	constructor(
		private appLoaderActions: AppLoaderActions,
		private route: ActivatedRoute
	) {}

	ngOnInit(): void {
		this.movies = this.route.snapshot.data.movies;
		this.appLoaderActions.stopLoading();
	}
}
