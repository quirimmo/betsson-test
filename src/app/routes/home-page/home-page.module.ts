import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HomePageComponent } from './home-page.component';
import { MoviesModule } from 'src/app/movies/movies.module';

@NgModule({
	declarations: [HomePageComponent],
	imports: [CommonModule, MoviesModule],
	providers: [HomePageComponent],
	bootstrap: []
})
export class HomePageRouteModule {}
