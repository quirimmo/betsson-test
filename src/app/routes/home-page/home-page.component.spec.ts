import { TestBed, async } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { HomePageComponent } from './home-page.component';
import { AppLoaderActions } from 'src/app/app-loader/app-loader.actions';
import { ActivatedRoute } from '@angular/router';
import Movie from 'src/app/movies/movie.model';
import { FilterMoviesPipe } from 'src/app/movies/pipes/filter-movies.pipe';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { MoviesListComponent } from 'src/app/movies/movies-list/movies-list.component';

const mockStopLoading = jest.fn();
const mockAppLoaderActions = {
	provide: AppLoaderActions,
	useValue: {
		stopLoading: mockStopLoading
	}
};
const mockActivatedRoute = {
	provide: ActivatedRoute,
	useValue: {
		snapshot: {
			data: { movies: ['movie 1'] }
		}
	}
};
const mockFilterMoviesPipe = {
	provide: FilterMoviesPipe,
	useValue: {
		transform: (movies: Movie[]) => movies
	}
};
const providers = [
	mockAppLoaderActions,
	mockActivatedRoute,
	mockFilterMoviesPipe
];
const imports = [RouterTestingModule, BrowserAnimationsModule];
const declarations = [HomePageComponent, MoviesListComponent];

const MockMoviesListComponent = {
	set: {
		inputs: ['movies'],
		template: '<span>movies list</span>'
	}
};

describe('HomePage Component', () => {
	let componentFixture, componentInstance;

	beforeEach(async(() => {
		TestBed.configureTestingModule({
			declarations,
			providers,
			imports
		})
			.overrideComponent(MoviesListComponent, MockMoviesListComponent)
			.compileComponents();
		componentFixture = TestBed.createComponent(HomePageComponent);
		componentInstance = componentFixture.debugElement.componentInstance;
	}));

	it('should create the component', () => {
		expect(componentFixture).toBeTruthy();
	});

	it('should match the snapshot', () => {
		expect(componentFixture).toMatchSnapshot();
	});

	describe('OnInit', () => {
		beforeEach(async () => {
			componentFixture.detectChanges();
			await componentFixture.whenStable();
		});
		it('should init the movies', () => {
			expect(componentInstance.movies).toEqual(['movie 1']);
		});

		it('should dispatch the stopLoading action', () => {
			expect(mockStopLoading).toHaveBeenCalled();
		});
	});
});
