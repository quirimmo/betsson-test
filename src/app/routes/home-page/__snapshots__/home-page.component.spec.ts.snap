// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Component should match the snapshot 1`] = `
<home-page
  appLoaderActions={[Function Object]}
  movies={[Function Array]}
  route={[Function Object]}
>
  <section
    class="container-fluid"
  >
    <movies-list
      class="ng-tns-c1-0"
    >
      <span
        class="ng-tns-c1-0"
      >
        movies list
      </span>
    </movies-list>
  </section>
</home-page>
`;
